lib/mongoid/atomic_update_preparer.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module Mongoid
+  # A singleton class to assist with preparing attributes for atomic
+  # updates.
+  #
+  # Once the deprecated Hash#__consolidate__ method is removed entirely,
+  # these methods may be moved into Mongoid::Contextual::Mongo as private
+  # methods.
+  #
+  # @api private
+  class AtomicUpdatePreparer
+    class << self
+      # Convert the key/values in the attributes into a hash of atomic updates.
+      # Non-operator keys are assumed to use $set operation.
+      #
+      # @param [ Class ] klass The model class.
+      # @param [ Hash ] attributes The attributes to convert.
+      #
+      # @return [ Hash ] The prepared atomic updates.
+      def prepare(attributes, klass)
+        attributes.each_pair.with_object({}) do |(key, value), atomic_updates|
```

Comment:
I believe you need to insert `key = klass.database_field_name(key)` immediately after this line. Please double check.

Reply:
Ah, took a minute, thought you were referring to something else. :) Right, I like that. I'll refactor.